lib/internal/loader/Loader.js

```diff
@@ -70,6 +71,15 @@ class Loader {
     const module = await job.run();
     return getNamespaceOfModuleWrap(module);
   }
+
+  static registerImportDynamicallyCallback(loader) {
+    setImportModuleDynamicallyCallback(async (referrer, specifier) => {
+      const normalizedBase = /^file:\/\//.test(referrer) ?
```

Comment:
This hack allows both CommonJS and ESM to use dynamic import and it's "guessing" which one it is. There's definitely cleaner ways to achieve this and this might break horribly in cases where the referrer is neither (e.g. `vm.runIn*Context`).

Reply:
I would add a `path.isAbsolute()` check to the non-URL code path just to make sure it is what we expect, and throw an error otherwise. `vm.runInContext()` + friends have a way to specify file path as well through the [`filename` option](https://nodejs.org/api/vm.html#vm_vm_runincontext_code_contextifiedsandbox_options).